python/foxglove-sdk/src/websocket_server.rs

```diff
@@ -0,0 +1,193 @@
+    #[new]
+    fn new(listener: Py<PyAny>) -> Self {
+        PyServerListener { listener }
+    }
+}
+
+/// A client connected to a running websocket server.
+#[pyclass(name = "Client", module = "foxglove")]
+pub struct PyClient {
+    #[pyo3(get)]
+    id: u32,
+}
+
+/// Information about a client channel.
+#[pyclass(name = "ClientChannelView", module = "foxglove")]
+pub struct PyClientChannelView {
+    #[pyo3(get)]
+    id: u32,
+    #[pyo3(get)]
+    topic: String,
+}
+
+/// Implementations of ServerListener which call the python methods. foxglove/__init__.py defines
+/// the `ServerListener` protocol for callers, since a `pyclass` cannot extend Python classes:
+/// https://github.com/PyO3/pyo3/issues/991
+///
+/// The ServerListener protocol implements all methods as no-ops by default; users extend this with
+/// desired functionality.
+///
+/// Methods on the listener interface do not return Results; any errors are logged, assuming the
+/// user has enabled logging.
+impl ServerListener for PyServerListener {
+    fn on_message_data(&self, client: Client, channel: ClientChannelView, payload: &[u8]) {
+        let client_info = PyClient {
+            id: client.id().into(),
+        };
+
+        let channel_view = PyClientChannelView {
+            id: channel.id().into(),
+            topic: channel.topic().to_string(),
+        };
+
+        let result: PyResult<()> = Python::with_gil(|py| {
+            let kwargs = PyDict::new(py);
+            kwargs.set_item("client", client_info)?;
+            kwargs.set_item("channel", channel_view)?;
+            kwargs.set_item("data", PyBytes::new(py, payload))?;
+
+            self.listener
+                .bind(py)
+                .call_method("on_message_data", (), Some(&kwargs))?;
+
+            Ok(())
+        });
+
+        if let Err(err) = result {
+            tracing::error!("Callback failed: {}", err.to_string());
+        }
+    }
+}
+
+/// Start a new Foxglove WebSocket server.
+///
+/// :param name: The name of the server.
+/// :param host: The host to bind to.
+/// :param port: The port to bind to.
+/// :param capabilities: A list of capabilities to advertise to clients.
+/// :param server_listener: A Python object that implements the :py:class:`ServerListener` protocol.
+/// :param supported_encodings: A list of encodings to advertise to clients.
+///    Foxglove currently supports "json", "ros1", and "cdr" for client-side publishing.
+///
+/// To connect to this server: open Foxglove, choose "Open a new connection", and select Foxglove
+/// WebSocket. The default connection string matches the defaults used by the SDK.
+#[pyfunction]
+#[pyo3(signature = (*, name = None, host="127.0.0.1", port=8765, capabilities=None, server_listener=None, supported_encodings=None))]
+pub fn start_server(
+    py: Python<'_>,
+    name: Option<String>,
+    host: &str,
+    port: u16,
+    capabilities: Option<Vec<PyCapability>>,
+    server_listener: Option<Py<PyAny>>,
```

Comment:
If we're taking a PyAny here, why do we need to export PyServerListener to Python? I don't think we need it all?

Reply:
Good catch, thanks. Leftover from before I realized it had to be implemented in Python directly.